### PR TITLE
Support for Batch Requests

### PR DESCRIPTION
**Pull request by dadheech115**
- Added classes to support Batch Requests as part of Graph Content.
- Added unit tests for these classes.

### CODE REVIEW

**Review by darrelmiller — APPROVED** [review]

File: MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+- (void)setUp {
+    // Put setup code here. This method is called before the invocation of each test method in the class.
+    [super setUp];
+}
+
+- (void)tearDown {
+    // Put teardown code here. This method is called after the invocation of each test method in the class.
+}
+
+- (void)testInit {
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m]: can you give it some other name, reflecting the test case?

File: MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    MSBatchRequestContent *requestContent = [[MSBatchRequestContent alloc] initWithRequests:nil error:&error];
+    XCTAssertNil(error);
+
+    MSBatchRequestStep *mockBatchStep = [[MSBatchRequestStep alloc] initWithId:@"1" request:self.requestForMock andDependsOn:nil];
+
+    requestContent  = [[MSBatchRequestContent alloc] initWithRequests:@[mockBatchStep] error:&error];
+    XCTAssertNil(error);
+
+    NSMutableArray *batchStepArray = [NSMutableArray new];
+    for(int i=0;i<21;i++) {
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m:45]: please put comment why 21 is significant here

File: MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    XCTAssertEqual(batchContentArray.count, 2);
+
+    //Test for duplicate request id
+    [requestContent addBatchRequestStep:driveBatchStep error:&error];
+    XCTAssertNotNil(error);
+    XCTAssertEqual(error.code, MSErrorCodeNonUniqueRequestId);
+    error = nil;
+
+    //Test for empty request id
+    MSBatchRequestStep *empty = [[MSBatchRequestStep alloc] initWithId:@"" request:self.requestForMock andDependsOn:nil];
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m:99]: can you create different tests for each test cases?

Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m:99]: Done.

File: MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m (hunk 2 above)
Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m:45]: Done.

File: MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m (hunk 1 above)
Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDKTests/GraphContent/BatchContent/MSBatchRequestContentTests.m]: Done.

File: MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+{
+    NSInteger maxNumberOfRequests;
+}
+
+- (instancetype)initWithRequests:(NSArray<MSBatchRequestStep *> *)batchRequestStepsArray error:(NSError *__autoreleasing *)error
+{
+    self = [super init];
+    if(self)
+    {
+        maxNumberOfRequests = 20;
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m:27]: can you put a comment that 20 is a enforcement by graph

File: MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        //Check whether number of request steps in array are more than the limit
+        if(batchRequestStepsArray.count>maxNumberOfRequests)
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m]: minor: space surrounding > operator?

File: MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            return;
+        }
+    }
+    [_batchRequestStepsArray addObject:requestStep];
+}
+
+- (void)removeBatchRequesStepWithId:(NSString *)requestId error:(NSError *__autoreleasing *)error
+{
+    BOOL didFindStep = NO;
+    for (long i = _batchRequestStepsArray.count-1 ; i >= 0 ; i--)
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m]: minor: no space before ; character?

File: MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        MSBatchRequestStep *requestStep = _batchRequestStepsArray[i];
+        for (long j = requestStep.arrayOfDependsOnIds.count-1 ; j >= 0 ; j--)
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m]: minor: no space before ; character?

**Review by deepak2016 — APPROVED** [review]

File: MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m (hunk 4 above)
Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m:27]: done.

File: MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m (hunk 5 above)
Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m]: done.

File: MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m (hunk 6 above)
Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m]: done.

File: MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m (hunk 7 above)
Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDK/GraphContent/BatchContent/MSBatchRequestContent.m]: done.

**Review by deepak2016 — APPROVED** [review]

**Pull request merged**